src/crush/CrushWrapper.cc

```diff
@@ -1230,6 +1230,46 @@ int CrushWrapper::remove_rule(int ruleno)
   return 0;
 }
 
+int CrushWrapper::update_device_class(CephContext *cct, int id, const string& class_name, const string& name)
+{
+  int class_id = get_class_id(class_name);
+  if (class_id < 0) {
+    ldout(cct, 0) << "update_device_class class " << class_name << " does not exist " << dendl;
+    return -ENOENT;
+  }
+  if (id < 0) {
+    ldout(cct, 0) << "update_device_class " << name << " id " << id << " is negative " << dendl;
+    return -EINVAL;
+  }
+  assert(item_exists(id));
+
+  if (class_map.count(id) != 0 && class_map[id] == class_id) {
+    ldout(cct, 5) << "update_device_class " << name << " already set to class " << class_name << dendl;
+    return 0;
+  }
+
+  if (class_map.count(id)) {
+    int old_class_id = class_map[id];
+    int class_id_count = 0;
+    for (auto &c : class_map)
+      if (c.second == old_class_id)
+	class_id_count++;
+    if (class_id_count < 2) {
+      ldout(cct, 0) << "update_device_class " << name << " id " << id << " is the last device "
+		    << "with class " << class_name[old_class_id] << ", manually edit the crushmap "
+		    << "to remove the class entirely" << dendl;
```

Comment:
This is the only bit that worries me.   If i set a class wrong (e.g. typo) i can't fix it.

Is the concern that you'll end up with an empty device tree?  (This seems fine to me.) Or no device tree at all? 

I'm forgetting how the tree generation works, but if we generate trees for any device class that is (1) used by an osd or (2) referenced by a rule that would prevent bad things from happening I think?